hfile/reader.go

```diff
@@ -205,27 +205,60 @@ func (r *Reader) FindBlock(from int, key []byte) int {
 }
 
 func (r *Reader) GetBlockBuf(i int, dst []byte) ([]byte, error) {
-	var err error
-
 	block := r.index[i]
 
 	switch r.CompressionCodec {
 	case CompressionNone:
 		dst = r.data[block.offset : block.offset+uint64(block.size)]
 	case CompressionSnappy:
-		uncompressedByteSize := binary.BigEndian.Uint32(r.data[block.offset : block.offset+4])
-		if uncompressedByteSize != block.size {
-			return nil, errors.New("mismatched uncompressed block size")
-		}
-		compressedByteSize := binary.BigEndian.Uint32(r.data[block.offset+4 : block.offset+8])
-		compressedBytes := r.data[block.offset+8 : block.offset+8+uint64(compressedByteSize)]
-		dst, err = snappy.Decode(dst, compressedBytes)
-		if err != nil {
-			return nil, err
+		/*
+		   SnappyCompressor internally uses BlockCompressorStream, which writes "blocks" of compressed data.
+
+		   These are NOT the "blocks" in the hfile sense. A "block" of an hfile may, when written by BlockCompressorStream
+		   write out many of what it calls "blocks". To avoid confusion, in this code, we shall call these "subblocks".
+
+		   Unfortunately the confusing usages of "block" do not stop there.
+
+		   BlockCompressorStream described the "blocks" (subblocks) that it writes thus[1]:
+		   "Each block contains the uncompressed length for the block, followed by one or more length-prefixed *blocks* of compressed data." (emphasis mine)
+
+		   Yes, that is a third, distinct, "block".
+
+		   In the hope that we might, with luck, navigate this correctly, we'll refer to these as "chunks" of "subblocks".
+
+		   Thus, reading a "block" of an hfile we read its "subblocks" in a loop, inside of which we, in a nested loop, read the subblock's chunks.
+
+		   1: http://grepcode.com/file/repo1.maven.org/maven2/org.apache.hadoop/hadoop-common/0.22.0/org/apache/hadoop/io/compress/BlockCompressorStream.java?av=f
```

Comment:
https://github.com/cloudera/hadoop-common/blob/4cd9f51a3f1ef748d45b8d77d0f211ad44296d4b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/io/compress/BlockCompressorStream.java is the exact version we're running i think